### DISCUSSION TITLE
package loading is very slow

### DISCUSSION THREAD

**Discussion by ishakdemir** [ALwrity/ALwrity#129] [General]
[image](https://private-user-images.githubusercontent.com/49456570/373855376-772b3f60-b200-42e9-ace6-71f1a32bbb1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgxMTU3NjYsIm5iZiI6MTcyODExNTQ2NiwicGF0aCI6Ii80OTQ1NjU3MC8zNzM4NTUzNzYtNzcyYjNmNjAtYjIwMC00MmU5LWFjZTYtNzFmMWEzMmJiYjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA1VDA4MDQyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllY2UxYjVhNTVjYWM1MTk5NjgyYTMwYTkzZTVjZDBlMWRiMzUzYWQyZTQyMjljYjNhNmQwMDc5M2M2MGE1ZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Kh_st_KwXMwQ0bNf9EI3tfDS25BYr0kWijMth0qdQz4)
Installing packages by doing pip install -r requirements.txt is progressing very slowly. The computer has been on for 2 hours.

**Comment by ishakdemir** [comment]
If it takes a long time and you encounter errors like me, you need to update pip
python -m pip install --upgrade pip

then such an error occurred

`File "D:\Programlar\python\lib\site-packages\streamlit\runtime\scriptrunner\exec_code.py", line 88, in exec_func_with_error_handling
    result = func()
File "D:\Programlar\python\lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 579, in code_to_exec
    exec(code, module.__dict__)
File "D:\AI-Writer\alwrity.py", line 9, in <module>
    from lib.utils.ui_setup import setup_ui
File "D:\AI-Writer\lib\utils\ui_setup.py", line 4, in <module>
    from .content_generators import content_planning_tools, ai_writers
File "D:\AI-Writer\lib\utils\content_generators.py", line 9, in <module>
    from lib.utils.seo_tools import ai_seo_tools
File "D:\AI-Writer\lib\utils\seo_tools.py", line 11, in <module>
    from lib.ai_marketing_tools.backlinking_ui_streamlit import backlinking_ui
File "D:\AI-Writer\lib\ai_marketing_tools\backlinking_ui_streamlit.py", line 3, in <module>
    from st_aggrid import AgGrid, GridOptionsBuilder, GridUpdateMode`

**Reply by ishakdemir** [reply]
requirements.txt -> streamlit-aggrid you need to add this too

**Comment by ishakdemir** [comment]
![image](https://github.com/user-attachments/assets/452d6d67-d3ad-4d4b-9e5f-22e964d38abd)

Do I have to enter the keys one by one in the code?

**Comment by AJaySi** [comment]
@ishakdemir 
let me fix this and update you today.

**Comment by AJaySi** [comment]
@ishakdemir 

I finally got around revamping requirements.txt and improving the loading performance of ALwrity. 
I also added lazy loading, cache and other performance optimizations. My local test seem to show improvements.
I tried multiple times and was able to install in minutes and not hours. 

Please help me validate the changes. I am adding more features, especially around DeepResearched Articles.

Regards.
